### PR TITLE
feat: add respective get_by_path UDFs for JSON type

### PR DESCRIPTION
**Pull request by CookiePieWw**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

#3686 #4515 #4230 

## What's changed and what's your intention?

Add respective get_by_path for json type. see https://github.com/datafuselabs/jsonb?tab=readme-ov-file#operators for the supported operators in the path.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

File: src/common/function/src/scalars/json/get_by_path.rs [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+                        columns.len() == 2,
+                        InvalidFuncArgsSnafu {
+                            err_msg: format!(
+                                "The length of the args is not correct, expect exactly two, have: {}",
+                                columns.len()
+                            ),
+                        }
+                    );
+                    let jsons = &columns[0];
+                    let strings = &columns[1];
```
Comment (WenyXu) [inline src/common/function/src/scalars/json/get_by_path.rs:86]: nit:
```suggestion
                    let paths = &columns[1];
```

File: src/common/function/src/scalars/json/get_by_path.rs [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+                    let size = jsons.len();
+                    let datatype = jsons.data_type();
+                    let mut results = [<$type VectorBuilder>]::with_capacity(size);
+
+                    match datatype {
+                        // JSON data type uses binary vector
+                        ConcreteDataType::Binary(_) => {
+                            for i in 0..size {
+                                let json = jsons.get_ref(i);
+                                let string = strings.get_ref(i);
```
Comment (WenyXu) [inline src/common/function/src/scalars/json/get_by_path.rs:97]: nit:
```suggestion
                                let path = strings.get_ref(i);
```

File: src/common/function/src/scalars/json/get_by_path.rs [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+                                let json = json.as_binary();
+                                let string = string.as_string();
```
Comment (WenyXu) [inline src/common/function/src/scalars/json/get_by_path.rs:100]:
```suggestion
                                let path = string.as_string();
```

File: src/common/function/src/scalars/json/get_by_path.rs [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+                                let result = match (json, string) {
+                                    (Ok(Some(json)), Ok(Some(string))) => {
+                                        get_json_by_path(json, string)
+                                            .and_then(|json| { jsonb::[<to_ $rust_type>](&json).ok() })
+                                    }
+                                    _ => None,
+                                };
```
Comment (WenyXu) [inline src/common/function/src/scalars/json/get_by_path.rs:107]:
```suggestion
                                let result = match (json, path) {
                                    (Ok(Some(json)), Ok(Some(path))) => {
                                        get_json_by_path(json, path)
                                            .and_then(|json| { jsonb::[<to_ $rust_type>](&json).ok() })
                                    }
                                    _ => None,
                                };
```

File: src/common/function/src/scalars/json/get_by_path.rs [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+                    let json = jsons.get_ref(i);
+                    let string = strings.get_ref(i);
+
+                    let json = json.as_binary();
+                    let string = string.as_string();
+                    let result = match (json, string) {
+                        (Ok(Some(json)), Ok(Some(string))) => get_json_by_path(json, string)
+                            .and_then(|json| jsonb::to_str(&json).ok()),
+                        _ => None,
+                    };
```
Comment (WenyXu) [inline src/common/function/src/scalars/json/get_by_path.rs:208]:
```suggestion
                for i in 0..size {
                    let json = jsons.get_ref(i);
                    let path = strings.get_ref(i);

                    let json = json.as_binary();
                    let path = string.as_string();
                    let result = match (json, path) {
                        (Ok(Some(json)), Ok(Some(path))) => get_json_by_path(json, path)
                            .and_then(|json| jsonb::to_str(&json).ok()),
                        _ => None,
                    };
```

File: tests/cases/standalone/common/function/json.sql
Comment (WenyXu) [inline tests/cases/standalone/common/function/json.sql:1]: Let's add more tests in future PRs. e.g., getting a field via the path with filter expr, getting the nested field from an array, etc.

**Review by WenyXu — APPROVED** [review]
Nice work! 🤩 Rest LGTM

File: tests/cases/standalone/common/function/json.sql
Comment (CookiePieWw) [reply tests/cases/standalone/common/function/json.sql:1]: Yes, sir! :wink:

**Comment by killme2008** [comment]
I really dislike the name "get_by_path_xxx" because it doesn't reflect any JSON-related meaning and could potentially conflict or be associated with other function names. Perhaps we could change it to "json_path_xxx"?

What do you think? @WenyXu @CookiePieWw

**Comment by WenyXu** [comment]
> I really dislike the name "get_by_path_xxx" because it doesn't reflect any JSON-related meaning and could potentially conflict or be associated with other function names. Perhaps we could change it to "json_path_xxx"?
> 
> What do you think? @WenyXu @CookiePieWw

How about `json_get_xxx`?👀

**Comment by killme2008** [comment]
> > I really dislike the name "get_by_path_xxx" because it doesn't reflect any JSON-related meaning and could potentially conflict or be associated with other function names. Perhaps we could change it to "json_path_xxx"?
> > What do you think? @WenyXu @CookiePieWw
> 
> How about `json_get_xxx`?👀

LGTM

**Comment by paomian** [comment]
LGTM, but we should add some test for this udf in the where clause in the next pr.

**Review by paomian — APPROVED** [review]

**Pull request merged**